File: backend/app/routers/suppliers.py
from datetime import datetime
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
import uuid

from app.db import get_db
from app.deps import get_current_user, get_tenant
from app.models import Tenant, User, Supplier, FinanceTransaction
from app.schemas import SupplierCreate, SupplierUpdate, SupplierOut
from app.services.finance_service import post_finance_transaction

router = APIRouter(prefix="/api/v1/ops/suppliers", tags=["suppliers"])


def _ensure_supplier_write_access(user: User):
    if str(user.role or "").lower() not in {"admin", "super_admin", "manager"}:
        raise HTTPException(status_code=403, detail="Manager access required")


@router.get("", response_model=list[SupplierOut])
def get_suppliers(
    db: Session = Depends(get_db),
    tenant: Tenant = Depends(get_tenant),
    user: User = Depends(get_current_user),
):
    _ensure_supplier_write_access(user)
    suppliers = db.query(Supplier).filter(Supplier.tenant_id == tenant.id).order_by(Supplier.name).all()
    return suppliers


@router.post("", response_model=SupplierOut)
def create_supplier(
    payload: SupplierCreate,
    db: Session = Depends(get_db),
    tenant: Tenant = Depends(get_tenant),
    user: User = Depends(get_current_user),
):
    _ensure_supplier_write_access(user)
    supplier = Supplier(
        id=str(uuid.uuid4()),
        tenant_id=tenant.id,
        name=payload.name,
        contact_person=payload.contact_person,
        phone=payload.phone,
        email=payload.email,
        address=payload.address,
        notes=payload.notes,
        balance=Decimal("0.00"),
        created_at=datetime.utcnow(),
    )
    db.add(supplier)
    db.commit()
    db.refresh(supplier)
    return supplier


@router.get("/{supplier_id}", response_model=SupplierOut)
def get_supplier(
    supplier_id: str,
    db: Session = Depends(get_db),
    tenant: Tenant = Depends(get_tenant),
    user: User = Depends(get_current_user),
):
    _ensure_supplier_write_access(user)
    supplier = db.query(Supplier).filter(Supplier.id == supplier_id, Supplier.tenant_id == tenant.id).first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")
    return supplier


@router.put("/{supplier_id}", response_model=SupplierOut)
def update_supplier(
    supplier_id: str,
    payload: SupplierUpdate,
    db: Session = Depends(get_db),
    tenant: Tenant = Depends(get_tenant),
    user: User = Depends(get_current_user),
):
    _ensure_supplier_write_access(user)
    supplier = db.query(Supplier).filter(Supplier.id == supplier_id, Supplier.tenant_id == tenant.id).first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")

    if payload.name is not None:
        supplier.name = payload.name
    if payload.contact_person is not None:
        supplier.contact_person = payload.contact_person
    if payload.phone is not None:
        supplier.phone = payload.phone
    if payload.email is not None:
        supplier.email = payload.email
    if payload.address is not None:
        supplier.address = payload.address
    if payload.notes is not None:
        supplier.notes = payload.notes

    db.commit()
    db.refresh(supplier)
    return supplier


@router.delete("/{supplier_id}")
def delete_supplier(
    supplier_id: str,
    db: Session = Depends(get_db),
    tenant: Tenant = Depends(get_tenant),
    user: User = Depends(get_current_user),
):
    _ensure_supplier_write_access(user)
    supplier = db.query(Supplier).filter(Supplier.id == supplier_id, Supplier.tenant_id == tenant.id).first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")

    db.delete(supplier)
    db.commit()
    return {"detail": "Supplier deleted successfully"}


class SupplierPaymentIn(BaseModel):
    amount: Decimal
    payment_source: str
    note: str | None = None


@router.post("/{supplier_id}/pay")
def pay_supplier(
    supplier_id: str,
    payload: SupplierPaymentIn,
    db: Session = Depends(get_db),
    tenant: Tenant = Depends(get_tenant),
    user: User = Depends(get_current_user),
):
    _ensure_supplier_write_access(user)
    supplier = db.query(Supplier).filter(Supplier.id == supplier_id, Supplier.tenant_id == tenant.id).first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")

    amount = Decimal(str(payload.amount)).quantize(Decimal("0.01"))
    if amount <= 0:
        raise HTTPException(status_code=400, detail="Payment amount must be > 0")

    source_code = str(payload.payment_source).strip().lower()
    if source_code not in {"cash", "card", "safe"}:
        raise HTTPException(status_code=400, detail="Payment source must be cash, card, or safe")

    post_finance_transaction(
        db,
        tenant_id=tenant.id,
        transaction_type="supplier_payment",
        amount=amount,
        source_code=source_code,
        destination_code="payable",
        created_by=user.username,
        category="Təchizatçı Ödənişi",
        counterparty=supplier.name,
        note=payload.note or f"{supplier.name} öhdəlik ödənişi",
        supplier_id=supplier.id,
    )

    supplier.balance -= amount
    db.commit()
    db.refresh(supplier)
    return {
        "id": supplier.id,
        "name": supplier.name,
        "balance": str(supplier.balance),
    }

File: src/api/suppliers.ts
import { v4 as uuidv4 } from 'uuid';
import { getDB, setDB } from '../lib/db_sim';
import { logEvent } from '../lib/logger';
import { apiRequest, isBackendEnabled } from './client';

export type Supplier = {
  id: string;
  tenant_id: string;
  name: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  address?: string;
  notes?: string;
  balance: string; // Stored as Decimal string
  created_at: string;
};

// Local storage simulation database helpers
const getSuppliersLocal = (tenantId: string): Supplier[] => {
  const tenantRows = getDB<Supplier>(`${tenantId}_suppliers`) || [];
  if (tenantRows.length > 0) return tenantRows;
  // Fallback to global db if not scoped
  const all = getDB<Supplier>('suppliers') || [];
  return all.filter((s) => String(s.tenant_id) === tenantId);
};

const saveSuppliersLocal = (tenantId: string, rows: Supplier[]) => {
  const shared = (getDB<Supplier>('suppliers') || []).filter((row) => String(row.tenant_id || '') !== tenantId);
  const next = [...shared, ...rows];
  setDB('suppliers', next);
  setDB(`${tenantId}_suppliers`, rows);
};

// Exported live functions
export async function get_suppliers_live(tenant_id: string): Promise<Supplier[]> {
  if (!isBackendEnabled()) {
    return Promise.resolve(getSuppliersLocal(tenant_id));
  }
  return apiRequest<Supplier[]>('/api/v1/ops/suppliers', { method: 'GET' });
}

export async function add_supplier_live(tenant_id: string, payload: {
  name: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  address?: string;
  notes?: string;
}): Promise<Supplier> {
  if (!isBackendEnabled()) {
    const nextList = getSuppliersLocal(tenant_id);
    const newSupplier: Supplier = {
      id: uuidv4(),
      tenant_id,
      name: payload.name,
      contact_person: payload.contact_person,
      phone: payload.phone,
      email: payload.email,
      address: payload.address,
      notes: payload.notes,
      balance: '0.00',
      created_at: new Date().toISOString(),
    };
    nextList.push(newSupplier);
    saveSuppliersLocal(tenant_id, nextList);
    logEvent('Admin', 'SUPPLIER_CREATED', { name: payload.name });
    return Promise.resolve(newSupplier);
  }
  return apiRequest<Supplier>('/api/v1/ops/suppliers', {
    method: 'POST',
    body: payload,
  });
}

export async function update_supplier_live(tenant_id: string, supplier_id: string, payload: {
  name?: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  address?: string;
  notes?: string;
}): Promise<Supplier> {
  if (!isBackendEnabled()) {
    const list = getSuppliersLocal(tenant_id);
    const index = list.findIndex((x) => x.id === supplier_id);
    if (index === -1) throw new Error('Supplier not found');
    const updated = { ...list[index], ...payload };
    list[index] = updated;
    saveSuppliersLocal(tenant_id, list);
    logEvent('Admin', 'SUPPLIER_UPDATED', { supplier_id });
    return Promise.resolve(updated);
  }
  return apiRequest<Supplier>(`/api/v1/ops/suppliers/${encodeURIComponent(supplier_id)}`, {
    method: 'PUT',
    body: payload,
  });
}

export async function delete_supplier_live(tenant_id: string, supplier_id: string): Promise<{ detail: string }> {
  if (!isBackendEnabled()) {
    const list = getSuppliersLocal(tenant_id);
    const nextList = list.filter((x) => x.id !== supplier_id);
    saveSuppliersLocal(tenant_id, nextList);
    logEvent('Admin', 'SUPPLIER_DELETED', { supplier_id });
    return Promise.resolve({ detail: 'Supplier deleted successfully' });
  }
  return apiRequest<{ detail: string }>(`/api/v1/ops/suppliers/${encodeURIComponent(supplier_id)}`, {
    method: 'DELETE',
  });
}

export async function pay_supplier_live(
  tenant_id: string,
  supplier_id: string,
  amount: number,
  payment_source: string,
  note?: string,
): Promise<{ id: string; name: string; balance: string }> {
  if (!isBackendEnabled()) {
    const list = getSuppliersLocal(tenant_id);
    const index = list.findIndex((x) => x.id === supplier_id);
    if (index === -1) throw new Error('Supplier not found');
    const supplier = list[index];
    const newBal = (parseFloat(supplier.balance) - amount).toFixed(2);
    supplier.balance = newBal;
    saveSuppliersLocal(tenant_id, list);
    logEvent('Admin', 'SUPPLIER_PAYMENT', { supplier_id, amount, payment_source, note });
    
    // Simulate updating finance accounts balance in local storage if necessary
    // (Local mock storage holds finance logs)
    const financeLogs = getDB<any>('finance_logs') || [];
    financeLogs.push({
      id: uuidv4(),
      tenant_id,
      transaction_type: 'supplier_payment',
      amount: amount.toFixed(2),
      source_code: payment_source,
      destination_code: 'payable',
      category: 'Təchizatçı Ödənişi',
      counterparty: supplier.name,
      note: note || `${supplier.name} öhdəlik ödənişi`,
      created_by: 'Admin',
      created_at: new Date().toISOString(),
    });
    setDB('finance_logs', financeLogs);

    return Promise.resolve({ id: supplier.id, name: supplier.name, balance: newBal });
  }
  return apiRequest<{ id: string; name: string; balance: string }>(
    `/api/v1/ops/suppliers/${encodeURIComponent(supplier_id)}/pay`,
    {
      method: 'POST',
      body: {
        amount,
        payment_source,
        note,
      },
    },
  );
}

// Helper to update local supplier balance during simulated restock (when backend is disabled)
export function adjust_local_supplier_balance_on_restock(tenant_id: string, supplier_id: string, total_price: number) {
  if (isBackendEnabled()) return;
  const list = getSuppliersLocal(tenant_id);
  const index = list.findIndex((x) => x.id === supplier_id);
  if (index !== -1) {
    const supplier = list[index];
    supplier.balance = (parseFloat(supplier.balance) + total_price).toFixed(2);
    saveSuppliersLocal(tenant_id, list);
  }
}
